### ISSUE TITLE
rectangular_symmetric yields unexpected output

### ISSUE THREAD

**Issue by lneuhaus** [XanaduAI/strawberryfields#193]
* *Expected behavior:* (What you expect to happen)
`sf.decompositions.rectangular_symmetric(np.identity(4)[::-1])`
should give a decomposition with all internal and external phases equal to zero, i.e.
```
([[0, 1, 0.0, 0.0, 4],
  [2, 3, 0.0, 0.0, 4],
  [1, 2, 0.0, 0.0, 4],
  [0, 1, 0.0, 0.0, 4],
  [2, 3, 0.0, 0.0, 4],
  [1, 2, 0.0, 0.0, 4]],
 array([0, 0, 0, 0]),
 None)
```
* *Actual behavior:* (What actually happens)
I get strange values, and negative phases.
```
([[0, 1, 0.0, 0.0, 4],
  [2, 3, 0.0, 0.0, 4],
  [1, 2, 0.0, -3.141592653589793, 4],
  [0, 1, 0.0, -4.71238898038469, 4],
  [2, 3, 0.0, 4.71238898038469, 4],
  [1, 2, 0.0, 4.71238898038469, 4]],
 array([1.8369702e-16+1.j, 3.0616170e-16+1.j, 6.1232340e-17+1.j,
        1.0000000e+00+0.j]),
 None)
```
The negative phases should be transformed into positive ones in the range `[0, 2*pi)`, and the values should be all zeros.
 
* *Reproduces how often:* (What percentage of the time does it reproduce?)
For any unitary above dimension 4 that I tried.

* *System information:* (post the output of `import stawberryfields as sf; sf.about()`)
```
Strawberry Fields: a Python library for continuous-variable quantum circuits.
Copyright 2018-2019 Xanadu Quantum Technologies Inc.

Python version:            3.6.8
Platform info:             Linux-4.15.0-55-generic-x86_64-with-debian-buster-sid
Installation path:         /home/leo/xanadu/code/strawberryfields/strawberryfields
Strawberry Fields version: 0.12.0-dev
Numpy version:             1.16.3
Scipy version:             1.2.1
The Walrus version:           0.7.0
Blackbird version:         0.2.1
TensorFlow version:        1.7.0
```

**Comment by lneuhaus** [comment]
I suspect the `numpy.fmod` in the implementation should simply be replaced by the regular modulus operator, and relevant tests should be added.

**Comment by josh146** [comment]
We should also add a test for the case `sf.decompositions.rectangular_symmetric(np.identity(4)[::-1])`!

**Comment by lneuhaus** [comment]
PR #196 makes all returned phases positive, but still results in nonzero (positive) external phase values for unitaries that can be obtained by composition of Mach-Zehner elements and final phase shifts with all phases equal to zero. For example (see below)for nine modes, this procedure gives the matrix `np.identity(9)[::-1], but the decomposition yields different values. Nevertheless, the test in `test_decompositions.py` that composes an original unitary with a recomposed decomposed version passes, meaning that both these decompositions are valid. Thus, the decomposition is not unique for certain settings, notably when `internal_phase` is 0 or pi. 

This can be understood when looking at one of the last MZI's in an interferometer that realizes a 'bar' or 'cross': its 'external' phase shift is directly imprinted onto one output mode, which goes straight through the Mach-Zehnder part without interfering with the other input mode, and eventually gets imprinted a final phase from the final phase shifter that is necessary to guarantee decomposability of arbitrary unitaries. 

Conclusion: it is unclear to me which decomposition should be chosen out of the infinite possibilities in this case. But the current decomposition, as it is in strawberry fields, is not wrong.

```import numpy as np
import strawberryfields as sf

internal_phase = 0.0
external_phase = 0.0
n_max = 9

print('composition')
unitary = np.identity(n_max)
for col in range(n_max):
    start = col % 2 # even column numbers start from mode 0, odd ones from 1
    for m in range(start, n_max-1, 2):
        ti = (m, m+1, internal_phase, external_phase, n_max)
        print(ti)
        mzi = sf.decompositions.mach_zehnder(*ti) 
        unitary = mzi @ unitary

print('composed unitary')
print(unitary)

print('decomposition')
tis, final_phase, _ = sf.decompositions.rectangular_symmetric(unitary)
print('Ti\'s')
for ti in tis:
    print(ti)
print('final_phases')
print(final_phase)
```

yiels

```
composition
(0, 1, 0.0, 0.0, 9)
(2, 3, 0.0, 0.0, 9)
(4, 5, 0.0, 0.0, 9)
(6, 7, 0.0, 0.0, 9)
(1, 2, 0.0, 0.0, 9)
(3, 4, 0.0, 0.0, 9)
(5, 6, 0.0, 0.0, 9)
(7, 8, 0.0, 0.0, 9)
(0, 1, 0.0, 0.0, 9)
(2, 3, 0.0, 0.0, 9)
(4, 5, 0.0, 0.0, 9)
(6, 7, 0.0, 0.0, 9)
(1, 2, 0.0, 0.0, 9)
(3, 4, 0.0, 0.0, 9)
(5, 6, 0.0, 0.0, 9)
(7, 8, 0.0, 0.0, 9)
(0, 1, 0.0, 0.0, 9)
(2, 3, 0.0, 0.0, 9)
(4, 5, 0.0, 0.0, 9)
(6, 7, 0.0, 0.0, 9)
(1, 2, 0.0, 0.0, 9)
(3, 4, 0.0, 0.0, 9)
(5, 6, 0.0, 0.0, 9)
(7, 8, 0.0, 0.0, 9)
(0, 1, 0.0, 0.0, 9)
(2, 3, 0.0, 0.0, 9)
(4, 5, 0.0, 0.0, 9)
(6, 7, 0.0, 0.0, 9)
(1, 2, 0.0, 0.0, 9)
(3, 4, 0.0, 0.0, 9)
(5, 6, 0.0, 0.0, 9)
(7, 8, 0.0, 0.0, 9)
(0, 1, 0.0, 0.0, 9)
(2, 3, 0.0, 0.0, 9)
(4, 5, 0.0, 0.0, 9)
(6, 7, 0.0, 0.0, 9)
composed unitary
[[0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 1.+0.j]
 [0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 1.+0.j 0.+0.j]
 [0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 1.+0.j 0.+0.j 0.+0.j]
 [0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 1.+0.j 0.+0.j 0.+0.j 0.+0.j]
 [0.+0.j 0.+0.j 0.+0.j 0.+0.j 1.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j]
 [0.+0.j 0.+0.j 0.+0.j 1.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j]
 [0.+0.j 0.+0.j 1.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j]
 [0.+0.j 1.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j]
 [1.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j]]
decomposition
Ti's
[0, 1, 0.0, 0.0, 9]
[2, 3, 0.0, 0.0, 9]
[1, 2, 0.0, 3.141592653589793, 9]
[0, 1, 0.0, 1.5707963267948966, 9]
[4, 5, 0.0, 0.0, 9]
[3, 4, 0.0, 3.141592653589793, 9]
[2, 3, 0.0, 3.141592653589793, 9]
[1, 2, 0.0, 3.141592653589793, 9]
[0, 1, 0.0, 1.5707963267948963, 9]
[6, 7, 0.0, 0.0, 9]
[5, 6, 0.0, 3.141592653589793, 9]
[4, 5, 0.0, 3.141592653589793, 9]
[3, 4, 0.0, 3.141592653589793, 9]
[2, 3, 0.0, 3.141592653589793, 9]
[1, 2, 0.0, 3.141592653589793, 9]
[0, 1, 0.0, 1.5707963267948966, 9]
[7, 8, 0.0, 1.5707963267948966, 9]
[6, 7, 0.0, 1.5707963267948966, 9]
[5, 6, 0.0, 1.5707963267948966, 9]
[4, 5, 0.0, 1.5707963267948972, 9]
[3, 4, 0.0, 1.5707963267948966, 9]
[2, 3, 0.0, 1.5707963267948966, 9]
[1, 2, 0.0, 1.5707963267948966, 9]
[0, 1, 0.0, 4.440892098500626e-16, 9]
[7, 8, 0.0, 4.71238898038469, 9]
[6, 7, 0.0, 4.71238898038469, 9]
[5, 6, 0.0, 4.71238898038469, 9]
[4, 5, 0.0, 4.71238898038469, 9]
[3, 4, 0.0, 4.71238898038469, 9]
[2, 3, 0.0, 4.71238898038469, 9]
[7, 8, 0.0, 4.71238898038469, 9]
[6, 7, 0.0, 4.71238898038469, 9]
[5, 6, 0.0, 4.71238898038469, 9]
[4, 5, 0.0, 4.71238898038469, 9]
[7, 8, 0.0, 4.71238898038469, 9]
[6, 7, 0.0, 4.71238898038469, 9]
final_phases
[ 1.0000000e+00-2.44929360e-16j  1.0000000e+00-4.44089210e-16j
 -6.2778623e-16-1.00000000e+00j -1.8369702e-16-1.00000000e+00j
 -1.0000000e+00+3.67394040e-16j -1.0000000e+00+1.22464680e-16j
  3.0616170e-16+1.00000000e+00j  6.1232340e-17+1.00000000e+00j
  1.0000000e+00+2.39393768e-16j]
```

**Comment by josh146** [comment]
@lneuhaus, what's the status of this issue?

**Comment by lneuhaus** [comment]
I will close this, because the current functionality in SF is not wrong, it just picks one out of many possible decompositions. A similar issue was fixed in #311.
